src/main/java/org/anarres/parallelgzip/ParallelGZIPOutputStream.java

```diff
@@ -71,42 +74,47 @@ public Block(@Nonnegative int index) {
          */
         // Only on worker thread
         @Override
-        public byte[] call() throws Exception {
+        public Block call() throws IOException {
             // LOG.info("Processing " + this + " on " + Thread.currentThread());
 
             State state = STATE.get();
             // ByteArrayOutputStream buf = new ByteArrayOutputStream(in.length);   // Overestimate output size required.
             // DeflaterOutputStream def = newDeflaterOutputStream(buf);
             state.def.reset();
             state.buf.reset();
-            state.str.write(in, 0, in_length);
+            state.str.write(buf, 0, bufLength);
             state.str.flush();
 
+            bufLength = state.buf.size();
+            assert bufLength <= buf.length;
+            state.buf.writeTo(buf);
+
             // return Arrays.copyOf(in, in_length);
-            return state.buf.toByteArray();
+            return this;
         }
 
         @Override
         public String toString() {
-            return "Block" /* + index */ + "(" + in_length + "/" + in.length + " bytes)";
+            return "Block" /* + index */ + "(" + bufLength + "/" + buf.length + " bytes)";
         }
     }
     // TODO: Share, daemonize.
     private final ExecutorService executor;
+    private final int nthreads;
     private final CRC32 crc = new CRC32();
-    private final int emitQueueSize;
-    private final BlockingQueue<Future<byte[]>> emitQueue;
-    private Block block = new Block(/* 0 */);
+    private final BlockingQueue<Future<Block>> emitQueue;
+    private ArrayList<Block> blocks = new ArrayList<>();  // list of reusable blocks
+    private Block block = new Block();
+    private byte[] blockBuf = block.buf;
     /** Used as a sentinel for 'closed'. */
     private long bytesWritten = 0;
 
     // Master thread only
     public ParallelGZIPOutputStream(@Nonnull OutputStream out, @Nonnull ExecutorService executor, @Nonnegative int nthreads) throws IOException {
         super(out);
         this.executor = executor;
-        // Some blocks compress faster than others; allow a long enough queue to keep all CPUs busy at least for a bit.
-        this.emitQueueSize = nthreads * 3;
-        this.emitQueue = new ArrayBlockingQueue<Future<byte[]>>(emitQueueSize);
+        this.nthreads = nthreads;
+        this.emitQueue = new ArrayBlockingQueue<Future<Block>>(nthreads);
```

Comment:
This still needs considerable beating up; I've asked a performance expert for help on this part, we may be able to profile it using JFC, which will help a lot.